asdf/converter.py

```diff
@@ -0,0 +1,192 @@
+        This method is not responsible for writing actual YAML; rather, it
+        converts an instance of a custom type to a tree of built-in Python types
+        (such as dict, list, str, or number), which can then be automatically
+        serialized to YAML as needed.
+
+        For container types returned by this method (dict or list),
+        the children of the container should not themselves be converted.
+        Any list elements or dict values will be converted by subsequent
+        calls to to_yaml_tree implementations.
+
+        The root node of the returned tree must be an instance of `dict`,
+        `list`, or `str`.  Descendants of the root node may be any
+        type supported by YAML serialization.
+
+        Parameters
+        ----------
+        obj : object
+            Instance of a custom type to be serialized.  Guaranteed to
+            be an instance of one of the types listed in the `types` class
+            attribute.
+
+        Returns
+        -------
+        dict or list or str
+            The tree representation of the object. Implementations that
+            wish to override the tag used to identify the object in YAML
+            are free to instead return an instance of `asdf.tagged.TaggedDict`,
+            `asdf.tagged.TaggedDict`, or `asdf.tagged.TaggedString`.
+        """
+        raise NotImplementedError("AsdfConverter subclasses must implement to_yaml_tree")
+
+    def from_yaml_tree(self, tree):
+        """
+        Convert a YAML subtree into an instance of a custom type.
+
+        For container types received by this method (dict or list),
+        the children of the container will have already been converted
+        by prior calls to from_yaml_tree implementations.
+
+        Note on circular references: trees that reference themselves
+        among their descendants must be handled with care.  Most
+        implementations need not concern themselves with this case, but
+        if the custom type supports circular references, then the
+        implementation of this method will need to return a generator.
+        Consult the documentation for more details.
+
+        Parameters
+        ----------
+        tree : dict or list or str
+            The YAML subtree to convert.  For the sake of performance, this
+            object will actually be an instance of `asdf.tagged.TaggedList`,
+            `asdf.tagged.TaggedDict`, or `asdf.tagged.TaggedString`.  These
+            objects should behave identically to their built-in counterparts.
+
+        Returns
+        -------
+        object
+            An instance of one of the types listed in the `types` class
+            attribute, or a generator that yields such an instance.
+        """
+
+        raise NotImplementedError("AsdfConverter subclasses must implement from_yaml_tree")
+
+
+# TODO(eslavich): Implement base class for converters that manipulate blocks
+
+
+class ConverterIndex:
```

Comment:
Is Index the best term for this, particularly in the Python context. First thing I think of is that it is going to give me a position in a list or array, whereas it is more like a book index. ConverterDirectory, ConverterWhitePages (sorry) or something more like that? It really provides two different modes of lookups

Reply:
ConverterCollection?  ConverterManager?  I took the Index name from the previous asdf class (TypeIndex) but I agree it could be better.

Reply:
ConverterCollection seems appropriate

Reply:
Should we poll to see what name everyone likes? Otherwise, this whole PR looks like a good basis to start from (after learning about secret RLock context manager behavior tricks and the like).

Reply:
`ConverterCollection` is good.

Reply:
`ConverterCollection` suits me fine too, I'll make that update